src/pg_query_parse_summary.c

```diff
@@ -0,0 +1,832 @@
+// and adds it to the state->from_clause_items list.
+void add_from_clause_item(Node *node, enum ContextType context, WalkState *state) {
+	if (!node)
+		return;
+
+	NodeWithContext* nwc = palloc(sizeof(NodeWithContext));
+	nwc->node = node;
+	nwc->context = context;
+
+	state->from_clause_items = lappend(state->from_clause_items, nwc);
+}
+
+// Given a List*-of-Nodes (clause) and WalkState (state), wrap them all
+// in NodeWithContext structs and add them to the state->from_clause_items list.
+void add_from_clause_items(List *clause, enum ContextType context, WalkState *state) {
+	if (!clause)
+		return;
+
+	ListCell *lc = NULL;
+	foreach(lc, clause) {
+		Node* node = lfirst(lc);
+		add_from_clause_item(node, context, state);
+	}
+}
+
+void add_function(List* funcname, enum ContextType context, WalkState* state){
+	StringInfoData buf;
+	initStringInfo(&buf);
+
+	bool first = true;
+	ListCell* lc = NULL;
+	foreach(lc, funcname) {
+		// Add a dot separator before everything but the first item.
+		if (first)
+			first = false;
+		else
+			appendStringInfoChar(&buf, '.');
+
+		String* str = (String*)lfirst(lc);
+		appendStringInfoString(&buf, quote_identifier(str->sval));
+	}
+
+	char* name = buf.data;
+	String* str_name = makeString(name);
+	Function* func = palloc(sizeof(Function));
+
+	func->name = str_name->sval;
+
+	// TODO: Verify this definition of schema_name is correct
+	if (list_length(funcname) > 1) {
+		String* first = linitial(funcname);
+		func->schema_name = first->sval;
+	} else {
+		func->schema_name = NULL;
+	}
+
+	// TODO: Verify this definition of function_name is correct.
+	String* last = llast(funcname);
+	func->function_name = last->sval;
+
+	func->context = context;
+
+	state->functions = lappend(state->functions, func);
+}
+
+void add_function_cstr(char* name, enum ContextType context, WalkState* state) {
+	Function* func = palloc(sizeof(Function));
+
+	func->name = name;
+	func->schema_name = NULL;
+	func->function_name = NULL;
+	func->context = context;
+
+	state->functions = lappend(state->functions, func);
+}
+
+// This should eventually be equivalent to https://github.com/pganalyze/pg_query/blob/main/lib/pg_query/parse.rb#L98-L340
+bool pg_query_parse_summary_walk_impl(Node* node, WalkState* state) {
+	if (node == NULL)
+		return false;
+
+	// NOTE: If `raw_expression_tree_walker()` doesn't handle a particular
+	// node type, you will need to `return false` early to avoid passing it
+	// to the tree walker.
+	// If you do this, be sure to update `pg_query_summary_stmt_walk_impl()`.
+
+	// NOTE: The Ruby implementation manually handles being given a List.
+	//
+	// In this version, the raw_expression_tree_walker() call at the
+	// end of this function will walk any items in a T_List node for us.
+
+	switch (nodeTag(node)) {
+		case T_CommonTableExpr: {
+			CommonTableExpr* cte = (CommonTableExpr*)node;
+			state->cte_names = lappend(state->cte_names, strdup(cte->ctename));
```

Comment:
```suggestion
			state->cte_names = lappend(state->cte_names, pstrdup(cte->ctename));
```

(generally, we should use `pstrdup` to utilize Postgres arena memory allocation, unless the result is returned to the caller, which should not be the case here since we encode all of this in a protobuf struct and intend to free this memory before returning the Protobuf to the caller)